### PR TITLE
"yarn knit" -- a command to test libraries while working on them

### PR DESCRIPTION
**Pull request by ide**
This is a proposal for a new command that is like `yarn link`/`npm link` and addresses several of the problems with `link`.

(This is an edited version of https://github.com/yarnpkg/yarn/issues/1213, written as an RFC as requested.)

### CODE REVIEW

File: text/0000-yarn-knit.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+# How We Teach This
+
+This proposal is mostly additive and affects only how people work on libraries that they are using in their apps. We would want to document the `knit` command in the "CLI Commands" section of the docs and perhaps add a new section to "The Yarn Workflow".
+
+`yarn link` would stay around, so people migrating from the npm client wouldn't have to learn anything new at first.
+
+# Drawbacks
+
+One issue with this proposal is that it's not clear what to put in the lockfile after running `yarn link dep` since we don't have an npm URL for the dep yet -- it hasn't been published to npm.
```
Comment (timdorr) [inline text/0000-yarn-knit.md]: What if `yarn knit` assigned a nonce version string and also spit out something like `To install in a local app, run: yarn knit dep@asdf1234`. That could be referenced in the lockfile with a `resolved` of `knit://dep@asdf1234`.

Comment (ide) [inline text/0000-yarn-knit.md]: I think that could work though one issue is that other team members wouldn't be able to resolve `knit://dep@asdf1234`. We could address that in a few ways:
1. Add a second yarn.local.lock file that is .gitignored and contains knitted dependencies and is overlaid on top of yarn.lock.
2. Commit knitted tarballs to Git. Yarn already supports offline installations.
3. Ignore it

**Comment by joshduck** [comment]
Symlinks have issues with peer dependencies. If love to see that accounted for here.  

Node resolves the link fully before walking up the path looking for node_modules. So if 'dep' has a peer depenency on React, it will not know that 'react' is actually app/node_modules/react.

**Comment by ide** [comment]
@joshduck Node has a flag called `--preserve-symlinks` which turns off link resolution. And the React packager and Webpack could consider a different resolution policy.

I'm not sure it's possible without symlinks to let people make changes in their library's directory and have the same changes automatically applied to their app's node_modules. There is a program called wml that watches for changes and copies files over but it relies on Watchman and doesn't work in some scenarios where symlinks would.

**Comment by bestander** [comment]
Sounds like a good idea

**Comment by bestander** [comment]
Considering that it is a brand new API, we might want to get feedback after we have a first implementation.
Any idea how we can implement it behind a flag maybe for a couple of releases?

**Comment by ide** [comment]
We could name the command "knit-EXPERIMENTAL" which has a similar effect as a flag without adding much complexity. As for getting people to test it out, Exponent is already using a copy of Yarn from upstream directly so it's relatively straightforward for us to use an unreleased version of Yarn.

**Comment by bestander** [comment]
Agree, great idea!

**Comment by louisgv** [comment]
Is this out yet?

**Comment by KyleAMathews** [comment]
I'd love to start developing Gatsby with this. It'd lower complexity getting started a lot with contributing to our Lerna monorepo setup.

**Comment by kristojorg** [comment]
Agreed, would love to see this happen

**Comment by bestander** [comment]
I would hold on implementing knit as described here, I think Yarn workspaces could cover all the usecases described here.

**Comment by bestander** [comment]
Workspaces should become a good end to end solution for monorepos, in [Workspaces phase 3 RFC](https://github.com/yarnpkg/rfcs/pull/66) we discuss linking between modules and it might be good enough to cover use cases raised here.

Thanks, @ide, for writing down this RFC, I got some ideas from it.

**Comment by sejoker** [comment]
After reading both RFCs on `yarn knit` and `workspaces linking` I tend to think they have different purpose and don't substitute one another. `yarn knit` is `yarn link` on steroids with a focus on the development experience. `Workspaces` on the other hand focus on the API while working in monorepo. I don't see how `Workspaces` address the main features of `yarn knit`:

- [Isolating node_modules correctly](https://github.com/yarnpkg/rfcs/blob/master/accepted/0000-yarn-knit.md#isolating-node_modules-correctly)
- [Working on multiple versions](https://github.com/yarnpkg/rfcs/blob/master/accepted/0000-yarn-knit.md#working-on-multiple-versions)

All in all, I find current RFC a good addition on top of `yarn link` even with rolled out `Workspaces` feature.

**Comment by saulshanabrook** [comment]
We had to write a custom linking mechanism in JupyterLab because if we used yarn link we ended up with duplicate packages in the webpack build. Basically we have a webpack build in one directory and then want to let users add their own folder (anywhere on the file system) and have that included in the webpack build and updated when the user updates their folder. We can't use workspaces for this since they are required to be subfolders of the root. And we can't use link, because then we end up with dependencies duplicated.

**Pull request merged**